✨ Add support for kube-vip

davidspek opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, minor or feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🏃 (:running:, other) -->

**What this PR does / why we need it**:
This replaces the binding of an elastic IP to one of the control plane nodes with deploying `kube-vip` and having it load balance the Kubernetes API between the various control plane nodes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

davidspek commented on templates/cluster-template-crs-cni.yaml:
```diff
@@ -47,14 +47,15 @@ spec:
-      - TRIMMED_KUBERNETES_VERSION=$(echo {{ .kubernetesVersion }} | sed 's/\./\\./g' | sed 's/^v//')
-      - RESOLVED_KUBERNETES_VERSION=$(apt-cache policy kubelet | awk -v VERSION=$${TRIMMED_KUBERNETES_VERSION} '$1~ VERSION { print $1 }' | head -n1)
-      - apt-get install -y ca-certificates socat jq ebtables apt-transport-https cloud-utils prips containerd kubelet=$${RESOLVED_KUBERNETES_VERSION} kubeadm=$${RESOLVED_KUBERNETES_VERSION} kubectl=$${RESOLVED_KUBERNETES_VERSION}
+      - |
```
This change is needed because `clusterctl generate cluster` will otherwise split some of the lines creating incorrect YAML which can cause some of the commands to not be executed.

cprivitere replied:
So I'm confused on this one, what about your changes make this necessary when it wasn't needed before?

davidspek replied:
@cprivite It actually is needed now as well, but it doesn't seem to have been causing major issues which is why it might not have been detected.

I came across this issue while I was trying to figure out why after a certain change to the commands the `kube-vip` command appeared to not have executed properly.

Without this change, the `preKubeadmCommands` YAML looks like this (using clusterctl 1.1.2 and 1.1.3):

![image](https://user-images.githubusercontent.com/28541758/157475911-5056a940-2eb6-4e86-b940-7431b3892d2e.png)

There are errors in the YAML formatting which causes some commands to not be executed.
- The `-` on line 64 should be part of the previous line
- Due to line 68 all following lines are seen as part of the same command
- If you manually fix line 68, line 70 will cause the same issue 

With this change, the formatting issues are avoided as you can see here:

![image](https://user-images.githubusercontent.com/28541758/157475507-e078b0e5-4295-4009-953c-5068fe9ba2e6.png)

cprivitere replied:
Gotcha, sorry about the confusion, I was checking a config generated with the old clusterctl and didn't realize this only happens with the newer one.

davidspek replied:
I didn't have an old version to check against, but it seemed like it needed to be fixed either way and shouldn't cause a backwards compatibility issue.

cprivite replied:
Are you testing in an environment that is using a version of kustomize later than 4.0.4? Versions after that have a bug that don't do the line wrapping well. That's why our make script is downloading 4.0.4 and not anything newer.

cprivitere replied:
I think I get the cause for this now, clusterctl is using a newer kustomize (or library associated with kustomize) inside of it that has issues with line wrapping long yaml lines. This is the same reason we use 4.0.4 and not anything newer in the Makefile. The upstream root cause is still an open issue sadly. :(

cprivitere replied:
On the plus side, with these changes we can just use the latest kustomize now in the makefile...I'll have to try that.

davidspek replied:
I was using Kustomize v4.0.5 so that would explain it. I just tested with the latest Kustomize v4.5.2 and the line wrapping issue is still present, but my workaround still works so that should indeed allow for using the latest Kustomize going forward. This is probably also a good way to prevent issues as users might have newer versions of Kustomize installed.

davidspek commented:
@cprivite I just pushed a commit that should fix the linting errors and I've validated the changes with a fresh cluster deployment.

davidspek commented on templates/cluster-template.yaml:
```diff
@@ -47,32 +47,37 @@ spec:
           export KUBECONFIG=/etc/kubernetes/admin.conf
           export CPEM_YAML=https://raw.githubusercontent.com/detiber/packet-ccm/test/deploy/template/deployment.yaml
-          export SECRET_DATA='cloud-sa.json=''{"apiKey": "{{ .apiKey }}","projectID": "${PROJECT_ID}", "eipTag": "cluster-api-provider-packet:cluster-id:${CLUSTER_NAME}"}'''
+          export SECRET_DATA='cloud-sa.json=''{"apiKey": "{{ .apiKey }}","projectID": "${PROJECT_ID}", "loadbalancer": "kube-vip://", "facility": "${FACILITY}"}'''
```
@cprivite I've noticed that for the CCM to function correctly and be able to assign IPs to services of type `LoadBalancer` that either the facility of metro needs to be added to the config. Since the facility is still being used by the cluster api provider, I've added the facility to the CCM config so it will work correctly for services as well.

The eipTag is removed since we don't want the CCM to allocate the EIP to one of the control plane nodes as `kube-vip` is already load balancing between the control plane nodes.

cprivite replied:
I believe @detiber is working on that particular issue right now. So we may not need this change.  Jason can you comment?

davidspek commented:
Would it make sense to bump the default number of control plane nodes to 3 as part of this PR? Since that is likely what most people should be using anyways.

cprivitere commented on templates/cluster-template-crs-cni.yaml:
```diff
@@ -140,18 +141,16 @@ spec:
-      EOF
-    - systemctl restart networking
+      if [ -f "/run/kubeadm/kubeadm-join-config.yaml" ]; then
+        KVVERSION=$(curl -sL https://api.github.com/repos/kube-vip/kube-vip/releases | jq -r ".[0].name")
```
Can you hard code the version of kube-vip? We want to test new kube-vip versions before we update cluster-api to start using them in the future.

davidspek replied:
Sure. Just simply hardcore it or should I expose it through and env var and have there be a default value?

davidspek replied:
I've just pinned the kube-vip version so I'm resolving this thread.

davidspek commented:
@detiber Do you think it would make sense to add this for the CAPI v1beta1 support?

detiber commented on templates/cluster-template.yaml:
```diff
@@ -47,32 +47,37 @@ spec:
-      - ping -c 3 -q {{ .controlPlaneEndpoint }} && echo OK || ip addr add {{ .controlPlaneEndpoint }} dev lo
+      - |
+        if [ -f "/run/kubeadm/kubeadm.yaml" ]; then
+          KVVERSION="v0.4.2"
+          ctr image pull ghcr.io/kube-vip/kube-vip:$${KVVERSION}
+          mkdir -p /etc/kubernetes/manifests/
+          ctr run --rm --net-host ghcr.io/kube-vip/kube-vip:$${KVVERSION} vip /kube-vip manifest pod --interface "lo" --vip "{{ .controlPlaneEndpoint }}" --controlplane --bgp --metal --metalKey "{{ .apiKey }}" --metalProjectID "b264ba74-f3bd-49f0-ae24-b0953765b7aa" | tee  /etc/kubernetes/manifests/kube-vip.yaml
```
This will expose the KUBE_VIP_VERSION as a variable to clusterctl to allow folks to easily override it if they wish.
```suggestion
          KVVERSION="${KUBE_VIP_VERSION:=v0.4.2}"
          ctr image pull ghcr.io/kube-vip/kube-vip:$${KVVERSION}
          mkdir -p /etc/kubernetes/manifests/
          ctr run --rm --net-host ghcr.io/kube-vip/kube-vip:$${KVVERSION} vip /kube-vip manifest pod --interface "lo" --vip "{{ .controlPlaneEndpoint }}" --controlplane --bgp --metal --metalKey "{{ .apiKey }}" --metalProjectID "b264ba74-f3bd-49f0-ae24-b0953765b7aa" | tee  /etc/kubernetes/manifests/kube-vip.yaml
```

detiber replied:
Should this use `bond0` for the interface rather than `lo`?

detiber replied:
It looks like projectID is being hardcoded here, it should use `${PROJECT_ID}` instead

davidspek replied:
The project ID is an accidental error in the template, I’ll fix that first thing in the morning along with making the kube-vip version configurable. 

As far as I’ve found in the kube-vip docs, the `lo` interface should be used:

[Note we bind the address to lo as we don't want multiple devices that have the same address on public interfaces.](https://kube-vip.chipzoller.dev/docs/installation/static/#bgp).

detiber replied:
Ah, gotcha, thanks for the clarification.

detiber commented on templates/cluster-template.yaml:
```diff
@@ -47,32 +47,37 @@ spec:
-          netmask 255.255.255.255
-        EOF
-      - systemctl restart networking
+        if [ -f "/run/kubeadm/kubeadm-join-config.yaml" ]; then
```
Are you handling join here separately to avoid `kubeadm join` from erroring during the pre-flight step? If so, you could configure it to ignore the pre-existing manifests directory.

When I was experimenting with this previously, I used this patch file to do just that: https://github.com/detiber/cluster-api-provider-packet/blob/capi-v0.5/templates/bare-with-kube-vip/kustomization.yaml#L6-L9

davidspek replied:
That’s indeed the reason. In general, I feel disabling checks that were created for a reason is a workaround that should be avoided if possible. In this case I think having the `kube-vip` pod manifest being created after the kubeadm join is probably the right way to go. The deployment has worked consistently for about 10 - 15 cluster deployments.

detiber replied:
I think my bigger concern here is potentially having drift between the commands run in the two separate places.

davidspek replied:
I’ll try and come up with a solution for possible drift. However, a similar issue also exists between the control plane and worker templates so a solution for this can also be useful there.

detiber commented on controllers/packetcluster_controller.go:
```diff
@@ -137,6 +137,11 @@ func (r *PacketClusterReconciler) reconcileNormal(ctx context.Context, clusterSc
 		}
 	}
 
+	if err := r.PacketClient.EnableProjectBGP(packetCluster.Spec.ProjectID); err != nil {
+		log.Error(err, "error enabling bgp for project")
+		return ctrl.Result{}, err
+	}
+
```
Is this needed because CPEM is not yet running when bootstrapping kube-vip?

davidspek replied:
CPEM actually isn’t needed for kube-vip to work. If that was the case, the bootstrap would fail. I just added this for general error handling.

detiber commented on controllers/packetmachine_controller.go:
```diff
@@ -362,6 +353,11 @@ func (r *PacketMachineReconciler) reconcile(ctx context.Context, machineScope *s
+	if err := r.PacketClient.EnsureNodeBGPEnabled(dev.ID); err != nil {
+		// Do not treat an error enabling bgp on machine as fatal
+		return ctrl.Result{RequeueAfter: time.Second * 20}, fmt.Errorf("failed to enable bpg on machine %s: %w", machineScope.Name(), err)
+	}
```
Is this needed because CPEM is not yet running when bootstrapping kube-vip?

davidspek replied:
CPEM actually isn’t needed for kube-vip to work. If that was the case, the bootstrap would fail. I just added this for general error handling.

detiber commented on pkg/cloud/packet/client.go:
```diff
@@ -38,6 +39,9 @@ const (
 	apiTokenVarName = "PACKET_API_KEY" //nolint:gosec
 	clientName      = "CAPP-v1beta1"
 	ipxeOS          = "custom_ipxe"
+	envVarLocalASN  = "METAL_LOCAL_ASN"
+	envVarBGPPass   = "METAL_BGP_PASS" //nolint:gosec
+	DefaultLocalASN = 65000
```
I'm wondering if it would make sense to add support to kube-vip for enabling BGP on the project and on the host, since it would be possible to introspect these values when running on a host.

I think requiring users to configure would be a bit of a pain, since it's not necessarily going to be the same information for all users or even for different facilities within Equinix Metal.

davidspek replied:
Kube-vip currently relies on CPEM to enable BGP on the project and nodes. The method I used here to determine/configure these values are derived from CPEM. So I’m not sure if introspection would work here since then CPEM should also be able to introspect these values.

detiber replied:
So, I fully agree that with the current scenario we have a bootstrapping problem between CPEM and kube-vip, what I am suggesting is that we could work with the kube-vip project (or contribute the changes needed ourselves) in order to support enabling BGP configuration where needed rather than doing it here.

davidspek replied:
I’d be happy to contribute the change to kube-vip. However, what I tried to get at is that it seems these values would need to be configured for kube-vip as well, as this is also the case for CPEM.

If I understand correctly, these values are later used by kube-vip to setup the BGP peering. So whatever enabled BGP on the project also needs to set the values, and there wouldn’t be a way to introspect them if BGP isn’t enabled on the project.

If what I’m saying doesn’t make sense I’m eager to understand why.

cprivitere replied:
I think the idea is that if you have kube-vip enable bgp on the project and the device, then the details like ASN and such can be pulled out via curl from the metadata server. 

https://metal.equinix.com/developers/docs/bgp/bgp-on-equinix-metal/

davidspek replied:
@cprivite That's true, but then wouldn't the same process that I'm doing here just move to `kube-vip`? I guess the main question is if that fits within the scope of what `kube-vip` does or is expected to do.

detiber reviewed (commented):
Overall, I think this is great work.

I'm wondering if it would make sense to not swap out the default cluster template, but rather to have an alternative template (cluster-template-kube-vip.yaml) that uses kube-vip vs CPEM for EIP management. This could even be extended to having a kube-vip template configured for bgp and a kube-vip template for EIP based configuration.

The main reason I bring this up is because I don't necessarily want to drop a major change to the default template on existing users and have it be a complete surprise for them. Especially if we don't have a plan for migrating existing clusters from CPEM EIP management to kube-vip bgp based management.

davidspek commented:
Thanks for the review and all your comments. I agree that it is a large change and the consequences that has for existing users. However, with the current changes to the code the CPEM EIP management wouldn’t work anymore. Part of me also believes  that `kube-vip` is the superior solution since it load balances requests tot the API and there should be minimal interruptions compared to having 2 standby control plane nodes. For that reason, I think it would be important for new users to use `kube-vip` for the control plane HA, and thus for it to be the default template.

If I can confirm the upgrade would be non-disruptive for current users and it is clearly stated in the upgrade/release docs, do you think that would be acceptable?

If not, I could try to make CPEM/kube-vip configurable and possibly make it so that a CPEM template is used existing users that upgrade. Otherwise, a separate template for `kube-vip` would still be an option.

detiber commented:
> Thanks for the review and all your comments. I agree that it is a large change and the consequences that has for existing users. However, with the current changes to the code the CPEM EIP management wouldn’t work anymore. Part of me also believes that `kube-vip` is the superior solution since it load balances requests tot the API and there should be minimal interruptions compared to having 2 standby control plane nodes. For that reason, I think it would be important for new users to use `kube-vip` for the control plane HA, and thus for it to be the default template.

Long term, I agree. However, considering existing users I don't want to have the default behavior change for them in a way that they wouldn't expect when spinning up a new cluster after the upgrade.

As a result, I think it would probably be better to keep the current default template, add a new template flavor (or multiple) for kube-vip and encourage users through documentation to start using the kube-vip template.

After we've been able to successfully migrate existing users over to kube-vip, then it would be easier to go ahead and change the default template without potentially causing users undue problems.

> 
> If I can confirm the upgrade would be non-disruptive for current users and it is clearly stated in the upgrade/release docs, do you think that would be acceptable?

I'm not sure there is a non-disruptive way to migrate users, since it would require changing out the KubeadmConfig and likely require a manual change to remove the EIP from the last host before kube-vip could properly configure BGP.

> 
> If not, I could try to make CPEM/kube-vip configurable and possibly make it so that a CPEM template is used existing users that upgrade. Otherwise, a separate template for `kube-vip` would still be an option.

davidspek commented:
I’ll test tomorrow and try and make the EIP handling configurable then.

In terms of upgrading existing users, as far as I know the controller will add a new control plane node, then remove an old one and repeat this until all have been upgraded. That’s the benefit of using reconciliation for managing cluster lifecycle. This shouldn’t cause interruptions to existing clusters, and also removes the problem of in assigning the EIP from the node, since the node will be removed.

davidspek commented:
@detiber Do you think configuring the EIP handling through an environment variable would be a sufficient solution?

cprivite commented:
Tried running this branch to create a cluster and got this error
```
E0311 21:43:59.423925      14 controller.go:317] controller/packetmachine "msg"="Reconciler error" "error"="failed to enable bpg on machine friday-control-plane-s8n9l: POST https://api.equinix.com/metal/v1/devices/a6b7f48e-7f05-4e35-8b15-e27d1457f3de/bgp/sessions: 422 Device has no assigned IP addresses " "name"="friday-control-plane-s8n9l" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="PacketMachine" 
```
It hung provisioning.

cprivitere commented:
Was testing this afternoon and it looks like for some reason kubevip isn't adding the bgp peer routes. I'm not sure why yet.

davidspek commented:
@cprivite That's strange. I have noticed it takes a while for the routes to show up in the Equinix Metal console, but the EIP for the control plane should come up pretty quickly.

cprivitere commented:
https://github.com/pluralsh/cluster-api-provider-packet/pull/1
I made a pull request to your branch to fix up a few things. Then I think it's ready for Jason to tell us how he'd like to have us package it up as a flavor template or whatever.

davidspek commented:
@cprivite Sorry for the delay here, had some other urgent things I needed to get done. I'll get back to this later today or tomorrow so it can move forward.

davidspek commented:
@detiber @cprivite  

I think there is one thing left for this to be ready for merging, based on what was discussed above.

- [ ] Make `kube-vip` optional

For that I would need to know how you would like to make it configurable. Do you think configuring the EIP handling through an environment variable would be a sufficient solution? There needs to be some way to enable or disable certain functions in the controller code.

cprivitere commented:
Yeah I'll be working on that now. Plan is to make it a second template that you'd choose by passing --flavor=kube-vip or something like that to clusterctl.

davidspek commented:
The extra template shouldn't be too difficult, it's mainly how to best pass that to the controller. I had already started with an implementation using an environment variable that the user would need to set.

detiber commented:
Apologies for causing the merge conflict on this PR, I did a first pass at rebasing/squashing my branch to clean things up for the upstream PR.

davidspek commented:
@detiber No worries. I'll do a rebase ASAP to fix up the merge conflicts.

davidspek commented:
@detiber Rebase should be good now

cprivitere commented:
I have pushed a PR to David's branch to split out the kube-vip support to a separate template so we don't change default behavior until some folks get a chance to use it in production. Once David approves it, the next steps that I see are:

- [ ] Readme updates incorporating kube-vip (essentially something that points out that --flavor kube-vip gets you a cluster where kube-vip manages the VIP EIP)
- [ ] Management of BGP on the project and device (making sure we have a proper workflow/documentation/expectations for users of what they need to do depending on which setup they use)
- [ ] Management of EIP in the controller (David and Jason, if you could elaborate on this one, I don't fully understand what we're missing here)

davidspek commented:
@cprivite I've just merged the PR. Regarding your second point, users shouldn't need to do any BGP related changes to the project or the devices as this should be handled by point 3. In this PR I added and removed some controller code related to BGP and the EIP. More precisely, I added code that enables BGP for the project and on the devices, and removed code that assigned the EIP to one of the control plane nodes. If both modes need to be supported, there needs to be a way to toggle some code in the controller on and off. I could do this by adding an environment variable that tells the controller that either kube-vip or the current EIP assigning is being used.

davidspek commented:
@detiber Would it make sense to also include support for using Metros instead of Facilities here?

cprivitere commented:
> @detiber Would it make sense to also include support for using Metros instead of Facilities here?

We'd like that to be a separate PR. We're very close to releasing the capiv1 update to upstream. After that I think we can do a rebase of this PR to release it against the new upstream, then we can do another PR for a feature like metros.

davidspek commented:
@cprivite Wouldn't it make sense to integrate this PR before release the CAPIv1 support?

cprivitere commented:
There are a number of questions and decisions around adding Metro support that we'd like to have it documented in an issue and have a place for people to easily find the decision process and submit feedback. 

https://github.com/kubernetes-sigs/cluster-api-provider-packet/issues/313

davidspek commented:
Sorry for not being clear. I meant the shouldn’t this PR for kube-vip support be merged so it is included in the CAPIv1 release.

cprivitere commented:
It would make sense, but there's some organization stuff happening (ie Jason is leaving Equinix Metal) that unfortunately means we need to keep it separate. The plan is to push the capiv1 update early next week and then have this PR re-submitted against the upstream kubernetes-sig/cluster-api-provider-packet repo.

detiber committed:
Publish images to multiple registries

- Also use quay.io by default

k8s-ci-robot committed:
Merge pull request #319 from detiber/multiRegistry

🏃🏻‍♂️ Publish images to multiple registries

cprivitere commented:
Ok, the public release of CAPIv1 support has released on https://github.com/kubernetes-sigs/cluster-api-provider-packet. Can you rebase this off the main branch and open a PR on the main repo? Then I can work on getting this released over there.

davidspek commented:
@cprivitere Sounds good. I'll move the PR over there.

davidspek committed:
fix hardcoded project id and make kube-vip version configurable

cprivitere committed:
run make generate

Signed-off-by: Chris Privitere <cprivite@users.noreply.github.com>

cprivitere committed:
remove caching of metadata where it isn't used

davidspek committed:
don't have CCM manage the control plane EIP

davidspek committed:
run make generate

davidspek committed:
run make generate

cprivitere committed:
split kube-vip version to separate template

davidspek committed:
remove comments from template

cprivitere committed:
Adding a space for readability.

Signed-off-by: Chris Privitere <cprivite@users.noreply.github.com>

cprivitere committed:
Clean up scripting, fix awk, add ip routes for IBX datacenters

Signed-off-by: Chris Privitere <cprivite@users.noreply.github.com>

cprivitere committed:
Update templates to set up BGP routes in IBX datacenters, update CPEM provider to released upstream version.

Signed-off-by: Chris Privitere <cprivite@users.noreply.github.com>

cprivitere committed:
Have kubeadm ignore manifests directory already existing.

Signed-off-by: Chris Privitere <cprivite@users.noreply.github.com>

cprivitere committed:
Remove echo used for debuggin

Signed-off-by: Chris Privitere <cprivite@users.noreply.github.com>

davidspek committed:
Add facility to CCM config so EIPs can be created for LoadBalancer services

davidspek committed:
fix linting errors

cprivitere committed:
Generate kube-vip template via kustomize

Signed-off-by: Chris Privitere <cprivite@users.noreply.github.com>

cprivitere committed:
Add ignore preflight errors to base template as it applies to all clusters < version 1.23, not just kube-vip ones.

Signed-off-by: Chris Privitere <cprivite@users.noreply.github.com>

davidspek committed:
pin kube-vip version

davidspek committed:
init

cprivitere committed:
Further refinements to the templates. bgp sections are only for kube-vip, remove systemctl restart networking to avoid networking service error.

Signed-off-by: Chris Privitere <cprivite@users.noreply.github.com>

cprivitere committed:
revert to non-kubevip state

davidspek committed:
make EIP management configurable

davidspek committed:
add suggested changes

Signed-off-by: DavidSpek <vanderspek.david@gmail.com>

davidspek committed:
run make generate

Signed-off-by: DavidSpek <vanderspek.david@gmail.com>

davidspek committed:
fix lint error

Signed-off-by: DavidSpek <vanderspek.david@gmail.com>

davidspek committed:
remove services from kube-vip config

Signed-off-by: DavidSpek <vanderspek.david@gmail.com>

cprivitere committed:
Convert to having the EIP_MANAGEMENT variable as part of the packetcluster Type (#5)

* Convert to having the EIP_MANAGEMENT variable as part of the packetcluster type

* Make vipmanager field immutable.

Signed-off-by: Chris Privitere <23177737+cprivitere@users.noreply.github.com>

* Rename field to VIPManager

Signed-off-by: Chris Privitere <23177737+cprivitere@users.noreply.github.com>

* rename to vipmanager, fix defaults, rm services

Signed-off-by: Chris Privitere <23177737+cprivitere@users.noreply.github.com>

* Fix typo and make VIPManager an enum

Signed-off-by: Chris Privitere <23177737+cprivitere@users.noreply.github.com>
